src/plugins/workspace/server/saved_objects/workspace_data_source_operation_wrapper.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getWorkspaceState } from '../../../../core/server/utils';
+import {
+  SavedObjectsBulkCreateObject,
+  SavedObjectsClientWrapperFactory,
+  SavedObjectsCreateOptions,
+  SavedObjectsUpdateOptions,
+  SavedObjectsFindOptions,
+  SavedObjectsDeleteOptions,
+  SavedObjectsBulkUpdateOptions,
+  SavedObjectsBulkUpdateObject,
+} from '../../../../core/server';
+import { DATA_SOURCE_SAVED_OBJECT_TYPE } from '../../../data_source/common';
+
+// Data source type saved object could not be created/updated/deleted in workspace exclude dashboard admin assign/unassign.
+// Assign would call update, unassign would call overwrite create or update.
+export class WorkspaceDataSourceOperationWrapper {
+  private isDataSourceType(type: SavedObjectsFindOptions['type']): boolean {
+    if (Array.isArray(type)) {
+      return type.every((item) => item === DATA_SOURCE_SAVED_OBJECT_TYPE);
+    }
+
+    return type === DATA_SOURCE_SAVED_OBJECT_TYPE;
+  }
+
+  private throwOperationError() {
+    throw new Error(
+      'You could not create/update/delete data source in workspace, please exit workspace and do it.'
+    );
+  }
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const isDashboardAdmin = getWorkspaceState(wrapperOptions.request)?.isDashboardAdmin;
+    const isInWorkspace = getWorkspaceState(wrapperOptions.request)?.requestWorkspaceId;
```

Comment:
What if the `workspace` is inside the options? I think this wrapper should be put behind id consumer and only consume the workspaces inside options.

Reply:
I would prefer to use `InWorkspace`  because what we want to limit is that user operations in the DSM inside workspace, it would look more directly. And this has more priority because id consumer would have some parse logic but this wrapper would limit directly based on role and operations, what do you think?

Reply:
I would say: the logic to get `workspaces` should be unified. I do not see there is any special use case for this wrapper to get workspaceId directly from workspaceState.

Imaging that a user calls
```
/w/${workspaceId}/updateObject, { options: { workspaces: null } }
```, should we consider it isInWorkspace or outOfWorkspace?

Reply:
Agree with @SuZhou-Joe, `options.workspaces` should be the source of truth, that's actually the value we are using to call saved object client to create the object.

Reply:
@raintygao Shall we restrict these data source operations when out of workspace? These client wrapper is for API level right? User can execute CUD operations if this API call doesn't include workspace info.

Reply:
Thanks @SuZhou-Joe @ruanyl , that makes sense, I would try to do some updates and take a look.

Reply:
@wanglam This PR is only aiming to limit data source operations in workspace, you could take a look at description of this PR.

Reply:
@raintygao What about creating a separate PR for the wrapper? It seems it's not related to default data source setting. This way, we can get default data source setting reviewed asap.

Reply:
That sounds good. I will do a separation.

Reply:
@ruanyl @SuZhou-Joe I have separated wrapper, there is no unresolved comment about default DS, could we merge it?